src/main/java/com/gregtechceu/gtceu/common/item/DataItemBehavior.java

```diff
@@ -81,27 +79,22 @@ public void appendHoverText(ItemStack stack, @Nullable Level level, List<Compone
     }
 
     @Override
-    public InteractionResult useOn(UseOnContext context) {
-        Level level = context.getLevel();
-        BlockPos pos = context.getClickedPos();
-        ItemStack stack = context.getItemInHand();
-        if (!level.isClientSide) {
-            MetaMachine machine = MetaMachine.getMachine(level, pos);
-            Pair<GTRecipeType, String> researchData = ResearchManager.readResearchId(stack);
-            if (machine instanceof MEPatternBufferPartMachine && researchData == null) {
-                stack.getOrCreateTag().putIntArray("pos", new int[] { pos.getX(), pos.getY(), pos.getZ() });
-            } else if (machine instanceof MEPatternBufferProxyPartMachine proxy) {
-                if (stack.hasTag()) {
-                    if (stack.getOrCreateTag().contains("pos", Tag.TAG_INT_ARRAY)) {
-                        int[] posArray = stack.getOrCreateTag().getIntArray("pos");
-                        BlockPos bufferPos = new BlockPos(posArray[0], posArray[1], posArray[2]);
-                        proxy.setBuffer(bufferPos);
-                    }
+    public InteractionResult onItemUseFirst(ItemStack itemStack, UseOnContext context) {
+        if (context.getLevel().getBlockEntity(context.getClickedPos()) instanceof MetaMachineBlockEntity blockEntity) {
+            if (blockEntity.getOwner() != null && !blockEntity.getOwner().isPlayerInTeam(context.getPlayer())) {
+                return InteractionResult.FAIL;
+            }
+            var machine = blockEntity.getMetaMachine();
+            if (machine instanceof IDataStickInteractable interactable) {
+                if (context.isSecondaryUseActive()) {
+                    return interactable.onDataStickShiftUse(context.getPlayer(), itemStack);
+                } else {
```

Comment:
Check to make sure there's no research on the data item before passing it to the machine to be written to.